server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -179,74 +186,78 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
     }
   }
 
-  public void addSegments(Iterable<DataSegment> segments, final DataSegmentChangeCallback callback)
+  private void addSegments(Collection<DataSegment> segments, final DataSegmentChangeCallback callback)
```

Comment:
You're saying this method should guarantee that the local cache has been completely loaded locally. I'm guessing this means explicitly that `serverManager.loadSegment` has been called for all the segments that were cached locally from a prior run of the service.

Ok, I get that.

At a cursory glance the BackgroundSegmentAnnouncer seems to be adding a bunch of extra complexity here. You have essentially 2 tasks per segment, one to load the segment and another to announce it. If I'm reading this correctly, the background announcements are on a delayed interval to do micro-batching of announcements as much as possible. Can that be done more cleanly in just this method?